File: adapters/medusa/compact.go
package main

import (
	"compress/gzip"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

type compactEvent struct {
	SchemaVersion int                      `json:"schema_version"`
	RecordType    string                   `json:"record_type"`
	Sequence      *sequenceRecord          `json:"sequence,omitempty"`
	Lineage       *lineageRecord           `json:"lineage,omitempty"`
	PrefixRefs    []compactPrefixReference `json:"prefix_references,omitempty"`
	NativePayload string                   `json:"native_payload_base64,omitempty"`
	SequenceCount int                      `json:"sequence_count"`
	PartialCount  int                      `json:"partial_count"`
}

type compactPrefixReference struct {
	MedusaHash   string `json:"medusa_hash"`
	IdentityHash string `json:"identity_hash"`
	StepCount    int    `json:"step_count"`
}

type compactEventWriter struct {
	path          string
	file          *os.File
	gzip          *gzip.Writer
	encoder       *json.Encoder
	sequenceCount int
	partialCount  int
	closed        bool
}

func newCompactEventWriter(path string) (*compactEventWriter, error) {
	if path == "" {
		return nil, fmt.Errorf("compact event output path is required")
	}
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return nil, err
	}
	file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
	if err != nil {
		return nil, err
	}
	compressed := gzip.NewWriter(file)
	return &compactEventWriter{
		path: path, file: file, gzip: compressed, encoder: json.NewEncoder(compressed),
	}, nil
}

func (writer *compactEventWriter) writeSequence(sequence *sequenceRecord, lineage *lineageRecord, refs []compactPrefixReference, payload []byte) error {
	if writer.closed {
		return fmt.Errorf("compact event stream is already closed")
	}
	if err := writer.encoder.Encode(compactEvent{
		SchemaVersion: 2, RecordType: "sequence", Sequence: sequence,
		Lineage: lineage, PrefixRefs: refs, NativePayload: base64.StdEncoding.EncodeToString(payload),
	}); err != nil {
		return err
	}
	writer.sequenceCount++
	return nil
}

func compactPrefixReferences(records []*sequenceRecord) ([]compactPrefixReference, error) {
	refs := make([]compactPrefixReference, 0, len(records))
	for _, record := range records {
		identity, err := stepSequenceIdentity(record.Steps)
		if err != nil {
			return nil, err
		}
		refs = append(refs, compactPrefixReference{
			MedusaHash: record.MedusaHash, IdentityHash: identity, StepCount: len(record.Steps),
		})
	}
	return refs, nil
}

func (writer *compactEventWriter) writePartial(sequence *sequenceRecord, refs []compactPrefixReference, payload []byte) error {
	if writer.closed {
		return fmt.Errorf("compact event stream is already closed")
	}
	if err := writer.encoder.Encode(compactEvent{
		SchemaVersion: 2, RecordType: "partial", Sequence: sequence,
		PrefixRefs: refs, NativePayload: base64.StdEncoding.EncodeToString(payload),
	}); err != nil {
		return err
	}
	writer.partialCount++
	return nil
}

func (writer *compactEventWriter) close() (map[string]any, error) {
	if writer.closed {
		return nil, fmt.Errorf("compact event stream is already closed")
	}
	writer.closed = true
	if err := writer.encoder.Encode(compactEvent{
		SchemaVersion: 2, RecordType: "eof", SequenceCount: writer.sequenceCount,
		PartialCount: writer.partialCount,
	}); err != nil {
		_ = writer.gzip.Close()
		_ = writer.file.Close()
		return nil, err
	}
	if err := writer.gzip.Close(); err != nil {
		_ = writer.file.Close()
		return nil, err
	}
	if err := writer.file.Sync(); err != nil {
		_ = writer.file.Close()
		return nil, err
	}
	if err := writer.file.Close(); err != nil {
		return nil, err
	}
	info, err := os.Stat(writer.path)
	if err != nil {
		return nil, err
	}
	digest, err := fileSHA(writer.path)
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"schema_version":   2,
		"path":             writer.path,
		"encoding":         "gzip-jsonl",
		"sequence_records": writer.sequenceCount,
		"partial_records":  writer.partialCount,
		"eof_record":       true,
		"compressed_bytes": info.Size(),
		"sha256":           digest,
	}, nil
}

File: src/seedbridge/compact.py
"""Fail-closed reader for Stage 3 gzip JSONL native evidence."""

from __future__ import annotations

import base64
import binascii
import gzip
import hashlib
import json
from pathlib import Path

from .io import file_hash
from .process import EvidenceMismatchError


def _error(message: str) -> EvidenceMismatchError:
    return EvidenceMismatchError(message, status="evidence_error")


def read_compact_events(path: Path, metadata: dict) -> dict:
    path = Path(path)
    if not path.is_file():
        raise _error("compact event stream is missing")
    if (metadata.get("schema_version") != 2
            or metadata.get("encoding") != "gzip-jsonl"
            or metadata.get("eof_record") is not True):
        raise _error("compact event metadata is incompatible")
    if metadata.get("sha256") != file_hash(path):
        raise _error("compact event checksum differs from its report")
    if metadata.get("compressed_bytes") != path.stat().st_size:
        raise _error("compact event byte count differs from its report")

    sequences: list[dict] = []
    lineage: list[dict] = []
    partial: list[dict] = []
    seen_indices: set[int] = set()
    known_prefix_hashes: set[str] = set()
    eof: dict | None = None
    try:
        with gzip.open(path, "rt", encoding="utf-8") as source:
            for line_number, line in enumerate(source, 1):
                if not line.strip():
                    raise _error(f"compact event stream has a blank line at {line_number}")
                try:
                    event = json.loads(line)
                except (TypeError, ValueError) as error:
                    raise _error(f"compact event line {line_number} is not JSON: {error}") from error
                if not isinstance(event, dict) or event.get("schema_version") != 2:
                    raise _error(f"compact event line {line_number} has an incompatible schema")
                if eof is not None:
                    raise _error("compact event stream contains records after EOF")
                kind = event.get("record_type")
                if kind == "eof":
                    eof = event
                    continue
                if kind not in {"sequence", "partial"}:
                    raise _error(f"compact event line {line_number} has an unknown record type")
                sequence = event.get("sequence")
                if not isinstance(sequence, dict) or type(sequence.get("sequence_index")) is not int:
                    raise _error(f"compact event line {line_number} has no sequence identity")
                index = sequence["sequence_index"]
                if index in seen_indices:
                    raise _error(f"compact event stream repeats sequence index {index}")
                seen_indices.add(index)
                encoded = event.get("native_payload_base64")
                try:
                    payload = base64.b64decode(encoded, validate=True)
                except (binascii.Error, TypeError, ValueError) as error:
                    raise _error(f"compact event {index} has invalid native payload encoding") from error
                if hashlib.sha256(payload).hexdigest() != sequence.get("native_digest"):
                    raise _error(f"compact event {index} native payload checksum differs")
                try:
                    native = json.loads(payload)
                except (UnicodeDecodeError, ValueError) as error:
                    raise _error(f"compact event {index} native payload is not JSON") from error
                if not isinstance(native, list) or not 1 <= len(native) <= 4:
                    raise _error(f"compact event {index} native payload is not a short call sequence")
                sequence["native_payload_verified"] = True
                references = event.get("prefix_references")
                if (not isinstance(references, list) or not references
                        or any(not isinstance(reference, dict) for reference in references)):
                    raise _error(f"compact event {index} has no prefix reference dictionary")
                reference_hashes = [reference.get("medusa_hash") for reference in references]
                if (any(not isinstance(value, str) or not value for value in reference_hashes)
                        or len(set(reference_hashes)) != len(reference_hashes)):
                    raise _error(f"compact event {index} has invalid prefix references")
                for position, reference in enumerate(references, 1):
                    if (type(reference.get("step_count")) is not int
                            or reference["step_count"] != position
                            or not isinstance(reference.get("identity_hash"), str)):
                        raise _error(f"compact event {index} has an invalid prefix reference")
                if kind == "partial":
                    if sequence.get("is_complete_sequence") is True or event.get("lineage") is not None:
                        raise _error(f"partial compact event {index} claims completion")
                    partial.append(sequence)
                    continue
                if sequence.get("is_complete_sequence") is not True:
                    raise _error(f"compact event {index} is not marked complete")
                event_lineage = event.get("lineage")
                if not isinstance(event_lineage, dict) or event_lineage.get("sequence_index") != index:
                    raise _error(f"compact event {index} has no matching lineage event")
                if (references[-1]["step_count"] != len(sequence.get("steps", []))
                        or references[-1]["medusa_hash"] != event_lineage.get("output_hash")
                        or references[-1]["identity_hash"]
                        != event_lineage.get("output_identity_hash")):
                    raise _error(f"compact event {index} final prefix differs from lineage")
                if event_lineage.get("kind") == "mutation":
                    parent_hashes = event_lineage.get("parent_hashes")
                    if (not isinstance(parent_hashes, list) or not parent_hashes
                            or any(parent not in known_prefix_hashes for parent in parent_hashes)):
                        raise _error(f"compact event {index} refers to an absent mutation parent")
                sequences.append(sequence)
                lineage.append(event_lineage)
                known_prefix_hashes.update(reference_hashes)
    except (OSError, EOFError, gzip.BadGzipFile) as error:
        raise _error(f"compact event stream is truncated or unreadable: {error}") from error

    if eof is None:
        raise _error("compact event stream has no EOF record")
    if eof.get("sequence_count") != len(sequences) or eof.get("partial_count", 0) != len(partial):
        raise _error("compact EOF counts differ from decoded records")
    if metadata.get("sequence_records") != len(sequences):
        raise _error("compact metadata sequence count differs")
    if metadata.get("partial_records") != len(partial):
        raise _error("compact metadata partial count differs")
    return {"sequences": sequences, "lineage": lineage, "partial_sequences": partial}
